ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java

```diff
@@ -41,30 +50,73 @@
     BeaconProposer proposer;
     BeaconChain beaconChain;
     Publisher publisher;
+    ValidatorConfig config;
 
     private ScheduledExecutorService proposerThread;
+    private ScheduledFuture currentTask;
+    private int validatorIdx;
+    private long lastStateRecalc;
 
-    public ProposerServiceImpl(BeaconProposer proposer, BeaconChain beaconChain, Publisher publisher) {
+    public ProposerServiceImpl(BeaconProposer proposer, BeaconChain beaconChain,
+                               Publisher publisher, ValidatorConfig config) {
         this.proposer = proposer;
         this.beaconChain = beaconChain;
         this.publisher = publisher;
+        this.config = config;
     }
 
     @Override
-    public void init() {
+    public void init(BeaconState state) {
+
+        Validator validator = state.getValidatorSet().getByPupKey(config.pubKey());
+        // something went wrong
+        if (validator == null) {
+            logger.error("Failed to start proposer: validator {} does not exist", Hex.toHexString(config.pubKey()));
+            return;
+        }
+
+        this.validatorIdx = validator.getIndex();
+
         this.proposerThread = Executors.newSingleThreadScheduledExecutor((r) -> {
             Thread t = Executors.defaultThreadFactory().newThread(r);
             t.setName("beacon-proposer-thread");
             t.setDaemon(true);
             return t;
         });
 
-        // trigger proposer each time when new block is imported
-        publisher.subscribe(BeaconBlockImported.class, data ->
-                this.submit(data.getBlock().getSlotNumber() + 1));
+        this.lastStateRecalc = state.getCrystallizedState().getLastStateRecalc();
+        // submit initial task
+        submitIfAssigned(state.getCommittees());
 
-        // schedule initial slot proposing
-        submit(proposer.getSlotNumber(System.currentTimeMillis()) + 2);
+        // listen to state updates
+        publisher.subscribe(BeaconBlockImported.class, (data) -> {
+            // trigger only if crystallized state has been recalculated
+            if (data.isBest() && data.getState().getCrystallizedState().getLastStateRecalc() > lastStateRecalc) {
+                this.lastStateRecalc = data.getState().getCrystallizedState().getLastStateRecalc();
+                this.submitIfAssigned(data.getState().getCommittees());
+            }
+        });
+    }
+
+    private void submitIfAssigned(Committee[][] committees) {
+        Committee.Index index = scanCommittees(validatorIdx, committees);
+        if (index.isEmpty())
+            return;
+
+        // validator from only the first committee is eligible to propose beacon chain block
+        if (index.getCommitteeIdx() > 0)
+            return;
+
+        // get number of the next slot that validator is eligible to propose
+        long slotNumber = calcNextProposingSlot(proposer.getSlotNumber(System.currentTimeMillis()),
```

Comment:
Let's move `System.currentTimeMillis()` somewhere to one source, as sharding logic is dependant of time and we may need some accuracy tweaks for it in future or whatever.